Flash layout extensions

mbolivar opened the pull request:
This builds upon work done by @nvlsianpu  and @nashif in https://github.com/zephyrproject-rtos/zephyr/pull/1562 by adding equivalent DTS properties and soc.h defines for STM32F4, STM32L4 and NXP K6X.

In addition, the last two patches:

1. select FLASH_PAGE_LAYOUT on nRF targets (2176afc)
2. extend the page layout API to add an iterator over pages (e9d0236)

These are useful to extend mcuboot to obtain the target configuration from Zephyr directly.

With link time garbage collection, 1. should add negligible size to the nRF flash driver when the page layout API is not used (flash storage for flash_nrf5_pages_layout()'s code and the pointer to it in the API struct, RAM for one struct flash_pages_layout, and a few extra instructions in nrf5_flash_init()). Hopefully that's acceptable, as it makes it much simpler to allow applications to use the API whenever it's available.

The iterator in 2. is useful for efficiently converting Zephyr's flash layout into what mcuboot expects at bootloader startup time.

Commit:
dts: nxp_k6x: add flash write-block-size

This makes the SoC flash compatible with the common nonvolatile flash
YAML schema, and provides a write alignment. It mirrors work done on
nRF chips for the DFU subsystem.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
dts: stm32f4: add flash write-block-size

This makes the SoC flash compatible with the common nonvolatile flash
YAML schema, and provides a write alignment. It mirrors work done on
nRF chips for the DFU subsystem.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
dts: stm32l4: add flash write-block-size

This makes the SoC flash compatible with the common nonvolatile flash
YAML schema, and provides a write alignment. It mirrors work done on
nRF chips for the DFU subsystem.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
arm: soc: nxp: k6x: add flash driver for IMG_MANAGER

Add definition required to enable the DFU subsystem image manager.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
arm: soc: stm32f4: add flash driver for IMG_MANAGER

Add definition required to enable the DFU subsystem image manager.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
arm: soc: stm32l4: add flash driver for IMG_MANAGER

Add definition required to enable the DFU subsystem image manager.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: flash: add flash_page_foreach()

This patch adds a routine which can be used to iterate over all flash
pages on the device.

This can be also done by using flash_get_page_info_by_idx(), but that
would add an unnecessary loop over the layout array for each page.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

nashif commented on drivers/flash/flash_page_layout.c:
```diff
@@ -80,3 +80,31 @@ size_t flash_get_page_count(struct device *dev)
 	return count;
 }
+
+void flash_page_foreach(struct device *dev, flash_page_cb cb, void *data)
```
need some code in Zephyr that uses this functions, preferably a testcase...

galak replied:
I would drop this from the PR until you have something that uses it.

mbolivar replied:
I'm using it out of tree, in mcuboot. I'll add a test case.

galak commented on arch/arm/soc/st_stm32/stm32f4/soc.h:
```diff
@@ -48,6 +48,12 @@
+/* For IMG_MANAGER */
+#if defined(CONFIG_SOC_FLASH_STM32)
+#define FLASH_DRIVER_NAME	CONFIG_SOC_FLASH_STM32_DEV_NAME
+#define FLASH_ALIGN		FLASH_WRITE_BLOCK_SIZE
```
where is FLASH_ALIGN used?

mbolivar replied:
In mcuboot.

mbolivar replied:
I added a patch which removes FLASH_ALIGN in Zephyr and just uses FLASH_WRITE_BLOCK_SIZE instead. This involved going back and taking it out of soc.h as was done by @nashif and @nvlsianpu in https://github.com/zephyrproject-rtos/zephyr/pull/1562. Hope that's OK.

mbolivar commented on drivers/flash/Kconfig:
```diff
@@ -127,6 +127,7 @@ config SOC_FLASH_NRF5
 	bool "Nordic Semiconductor nRF5X flash driver"
 	depends on FLASH && SOC_FAMILY_NRF5
 	default n
+	select FLASH_PAGE_LAYOUT
```
Okay, that works too.

mbolivar replied:
I've dropped this patch.

Commit:
samples: flash_shell: update to keep up with flash driver changes

Keep the flash shell up to date with the latest flash driver updates.

- Get the driver name from soc.h
- Add a write_block_size command
- Implement flash_shell_page_layout() using flash_page_foreach()

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
dfu: replace FLASH_ALIGN with FLASH_WRITE_BLOCK_SIZE

Use the define generated by the DTS instead of using the FLASH_ALIGN
alias. The latter is an internal mcuboot name. We shouldn't need it in
Zephyr itself.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

mbolivar commented:
Hi, I believe I've addressed open comments.

galak commented:
why aren't we getting the IMG_MANAGER stuff from dt instead.

mbolivar commented:
> why aren't we getting the IMG_MANAGER stuff from dt instead.

Sorry, I missed this comment.

Are you saying you want the defines, like the one in https://github.com/zephyrproject-rtos/zephyr/pull/1594/commits/8342256a1ecd69bbc328b343d399ca609067169a, to come from device tree instead?

I.e. we should remove CONFIG_SOC_FLASH_some_soc_NAME from Kconfig and put it in DT instead? Perhaps as a compatible string? Sorry for the confusion.

mbolivar commented:
> I.e. we should remove CONFIG_SOC_FLASH_some_soc_NAME from Kconfig and put it in DT instead? Perhaps as a compatible string? Sorry for the confusion.

@galak this question is for you

nashif commented:
I guess we can take it as is right now and see how to make this better work with DT in a followup PR.
